evm_loader/program/src/instruction/transaction.rs

```diff
@@ -141,20 +146,44 @@ fn finalize<'a>(
     mut storage: State<'a>,
     account_storage: &mut ProgramAccountStorage<'a>,
     results: Option<EvmResults>,
-    used_gas: U256,
+    mut used_gas: U256,
+    mut gasometer: Gasometer,
+    accounts_operations: Option<AccountsOperations>,
 ) -> ProgramResult {
     debug_print!("finalize");
 
-    // The only place where checked math is requiered.
-    // Saturating math should be used everywhere else for gas calculation
+    let accounts_operations = accounts_operations.map_or_else(
+        || results.as_ref().map_or_else(
+            HashMap::default,
+            |(_, _, actions)| {
+                let accounts_operations = account_storage.calc_acc_changes(actions);
+                gasometer.record_accounts_operations(&accounts_operations);
+                used_gas = gasometer.used_gas();
+
+                accounts_operations
+            }
+        ),
+        |accounts_operations| accounts_operations,
+    );
+
+    // The only place where checked math is required.
+    // Saturating math should be used everywhere else for gas calculation.
     let total_used_gas = storage.gas_used.checked_add(used_gas);
 
     // Integer overflow or more than gas_limit. Consume remaining gas and revert transaction with Out of Gas
-    if total_used_gas.is_none() || (total_used_gas > Some(storage.gas_limit))  {
+    if total_used_gas.is_none() || (total_used_gas > Some(storage.gas_limit)) {
         let out_of_gas = Some((vec![], ExitError::OutOfGas.into(), None));
         let remaining_gas = storage.gas_limit.saturating_sub(storage.gas_used);
 
-        return finalize(accounts, storage, account_storage, out_of_gas, remaining_gas);
+        return finalize(
+            accounts,
+            storage,
+            account_storage,
+            out_of_gas,
+            remaining_gas,
+            gasometer,
+            Some(accounts_operations),

```

Comment:
Bug here. You commit `out_of_gas` error in transaction receipt, but `accounts_operations` contains actions for success execution. This lead to creation accounts on error and ineffective spends operator's lamports.

In case of inner call `finalize` you need to rebuild `accounts_operations` from actial `actions` (which empty).
So you can remove `accounts_operations` from parameters of `finalize` function and calculate it in place (and simplify logic from 155-167 lines!)

See also this case in `transaction_execute_from_instruction`.

Reply:
Done.